monix-tail/shared/src/main/scala/monix/tail/internal/IterantDump.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2014-2017 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.tail.internal
+
+import java.io.PrintStream
+
+import cats.effect.Sync
+import cats.syntax.all._
+import monix.execution.misc.NonFatal
+import monix.tail.Iterant
+import monix.tail.Iterant.{Halt, Last, Next, NextBatch, NextCursor, Suspend}
+import monix.tail.internal.IterantUtils.signalError
+
+private[tail] object IterantDump {
+  /**
+    * Implementation for `Iterant#dump`
+    */
+  def apply[F[_], A](source: Iterant[F, A], prefix: String, out: PrintStream = System.out)
+                    (implicit F: Sync[F]): Iterant[F, A] = {
+    var pos = 0
```

Comment:
I don't like this `var`, for one because it breaks referential transparency even when the `source` is a `Suspend`.

Think of referential transparency as the ability to execute the thing multiple times and get the same result. By closing over this `var` you'll get different results on different runs, as you have shared mutable state.

You don't really need a `var` with your `loop`, when you can simply do:

```scala
def loop(pos: Long)(source: Iterant[F, A]): Iterant[F, A]
```